Handle error for create_partition

fosterseth opened the pull request:
<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: "Fixed 'already exists' error when to creating a table partition"
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
https://github.com/ansible/awx/issues/11811

`create_partition` should already skip creating a table if it already exists.

```python3
with connection.cursor() as cursor:
    cursor.execute(
        f'CREATE TABLE IF NOT EXISTS {tblname}_{partition_label} '
        f'PARTITION OF {tblname} '
        f'FOR VALUES FROM (\'{start_timestamp}\') to (\'{end_timestamp}\');'
    )
```
However, occasionally parallel operations collide and we get an error.



I can recreate the scenario with

```python3
from awx.main.utils.common import create_partition
from django.utils import timezone
from multiprocessing import Pool

def x(i):
    create_partition('main_projectupdateevent', start=timezone.now())

with Pool(30) as p:
    print(p.map(x, range(30)))
```
 
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 20.0.2.dev121+g26c183b870

```

AlanCoding reviewed (changes requested):
There's 1 more call to this method in `tasks/jobs.py`, and even though it's already in a try-except, that does not accomplish the intention of this. Following the current pattern, I think you should add another _nested_ try-except in that case, so that an error here does not prevent the inventory update from running.

Or you could put this in the `create_partition` method itself. Up to you.

jbradberry reviewed (changes requested):
I think I have to throw a red flag here.  A failed write to the database here will likely cause further attempts to write to the database in completely unrelated code to also fail.  I strongly suspect that we need transactions and rollbacks to do this right.

AlanCoding commented:
transactions have their own way of making error handing worse, but Seth's script looks like a good test for the solution, whatever that may be.

fosterseth committed:
Handle error for create_partition

Occasionally the create_partition will error with,
relation "main_projectupdateevent_20220323_19" already exists

This change wraps the db command into a try except block with its
own transaction

fosterseth commented on awx/main/utils/common.py:
```diff
@@ -1141,12 +1141,16 @@ def create_partition(tblname, start=None, end=None, partition_label=None, minute
+                    f'FOR VALUES FROM (\'{start_timestamp}\') to (\'{end_timestamp}\');'
+                )
+    except Exception as e:
+        logger.warning(e)
```
not sure of the best log level here -- exception and error is going to show a bunch of scary red text, and we probably don't want users to panic or open issues if they see it.

I think warning is okay-ish, but I could be convinced of debug or even just passing with no logging.

jbradberry replied:
Since the situation is not actually very exceptional, I'd go for no higher than `info`, possibly `debug`.

AlanCoding replied:
`info` or `debug` would be just fine if you were sure that you were catching the error you think you're catching. With a blanket `except Exception`, something could have gone wrong that you didn't expect. If you could narrow the error type, then a low-level log would be appropriate. If you can't do that, then you should use `logger.exception`

fosterseth replied:
The "already exists" error is a ProgrammingError, which seems narrow enough to catch on

https://peps.python.org/pep-0249/#programmingerror

```
Exception raised for programming errors, e.g. table not found or already exists, syntax error in the SQL statement, wrong number of parameters specified, etc. It must be a subclass of [DatabaseError](https://peps.python.org/pep-0249/#databaseerror).
```

I'm thinking we should just catch the ProgrammingError and let all other errors raise an exception.

fosterseth commented:
with django's autocommit mode, we might not even need a `with transactions.atomic()` wrapper around the cursor.execute() call.

```
try:
    cursor.execute(..)
except:
    pass
```

this would work under normal conditions. If cursor.execute hits an error, it will be rolled back automatically and we can carry on with subsequent queries.

However, if we are already in a transaction block by the time we hit this code (i.e. a `with transaction.atomic():` higher in the call stack), then this try / except block isn't enough. Once we hit the error, nothing is rolled back, and then try to run more queries in the same thread. This would lead to the "An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block." error.

So for that, wrapping the transaction.atomic() is safer, or we could consider adding the `durable=True` flag to ensure it's the outer transaction block.

#### Demonstration

To demonstrate that we can make queries right after the errors,

```python3
from awx.main.utils.common import create_partition
from django.utils import timezone
from multiprocessing import Pool

def x(i):
  try:
    create_partition('main_projectupdateevent', start=timezone.now())
  except:
    pass
  return Job.objects.count()

def y(i):
  with transaction.atomic():
    return x(i)

with Pool(30) as p:
  print(p.map(y, range(30)))

```

Note: the outer transaction.atomic() is necessary to demonstrate the error. On devel, calling create_partition() will run the cursor.execute() in its own transaction (and thus will auto rollback), so we would never see the "ignored until end of transaction block" error.

##### output on current devel branch:

`InternalError: current transaction is aborted, commands ignored until end of transaction block`

##### on this branch:

```
2022-03-28 17:14:29,236 WARNING  [-] awx.main.utils relation "main_projectupdateevent_20220328_17" already exists
..
2022-03-28 17:14:29,296 WARNING  [-] awx.main.utils relation "main_projectupdateevent_20220328_17" already exists
2022-03-28 17:14:29,294 WARNING  [-] awx.main.utils relation "main_projectupdateevent_20220328_17" already exists
2022-03-28 17:14:29,300 WARNING  [-] awx.main.utils relation "main_projectupdateevent_20220328_17" already exists
[51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51]
```

51 is the output from `Job.objects.count()`

AlanCoding commented on awx/main/utils/common.py:
```diff
@@ -1141,12 +1141,16 @@ def create_partition(tblname, start=None, end=None, partition_label=None, minute
+                    f'FOR VALUES FROM (\'{start_timestamp}\') to (\'{end_timestamp}\');'
+                )
+    except ProgrammingError as e:
+        logger.debug(e)
```
A little more context would be appreciated, because you don't get line numbers or anything. If a log statement looks absurdly redundant, you're doing it right. Like `f'Caught known error due to existing partition: {e}'`

AlanCoding reviewed (approved):
great research on this

rebeccahhh reviewed (approved):
Loving the research and changes. +1

fosterseth commented:
thanks team, going to make sure tests pass, then I'll merge it in :)
